crates/biome_js_analyze/src/analyzers/nursery/no_unused_private_class_members.rs

```diff
@@ -0,0 +1,299 @@
+    type State = RuleState;
+    type Signals = Vec<Self::State>;
+    type Options = ();
+
+    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
+        let node = ctx.query();
+        let private_members: FxHashMap<TokenText, RuleState> =
+            get_all_declarated_private_members(node);
+
+        if private_members.is_empty() {
+            return vec![];
+        }
+
+        traverse_members_usage(node.syntax(), private_members)
+    }
+
+    fn diagnostic(_: &RuleContext<Self>, state: &Self::State) -> Option<RuleDiagnostic> {
+        Some(RuleDiagnostic::new(
+            rule_category!(),
+            state.range,
+            markup! {
+                "This private class member is defined but never used."
+            },
+        ))
+    }
+
+    fn action(ctx: &RuleContext<Self>, state: &Self::State) -> Option<JsRuleAction> {
+        let mut mutation = ctx.root().begin();
+
+        mutation.remove_node(state.member.clone());
+
+        Some(JsRuleAction {
+            category: ActionCategory::QuickFix,
+            applicability: Applicability::MaybeIncorrect,
+            message: markup! { "Remove unused declaration." }.to_owned(),
+            mutation,
+        })
+    }
+}
+
+/// Check for private member usage
+/// if the member usage is found, we remove it from the hashmap
+fn traverse_members_usage(
+    syntax: &SyntaxNode<JsLanguage>,
+    mut private_members: FxHashMap<TokenText, RuleState>,
+) -> Vec<RuleState> {
+    let iter = syntax.preorder();
+
+    for walk in iter {
+        match walk {
+            biome_rowan::WalkEvent::Enter(node) => {
+                if let Some(js_name) = AnyJsName::cast_ref(&node) {
+                    let value_token = match &js_name {
+                        AnyJsName::JsName(name) => name.value_token(),
+                        AnyJsName::JsPrivateName(name) => name.value_token(),
+                    };
+
+                    if let Ok(value_token) = value_token {
+                        let token = value_token.token_text_trimmed();
+
+                        if let Some(state) = private_members.get(&token) {
+                            let is_accessor = matches!(
+                                state.member.syntax().kind(),
+                                JsSyntaxKind::JS_SETTER_CLASS_MEMBER
+                                    | JsSyntaxKind::JS_GETTER_CLASS_MEMBER
+                            );
+
+                            if is_write_only(&js_name) && !is_accessor {
+                                continue;
+                            }
+
+                            if is_update_expression(&js_name) {
+                                continue;
+                            }
+                        }
+
+                        private_members.remove(&token);
+                    }
+                }
+            }
+            biome_rowan::WalkEvent::Leave(_) => {
+                if private_members.is_empty() {
+                    break;
+                }
```

Comment:
We could move this statement just after `private_members.remov`.

In this case we could write:

```rs
for node in syntax.descendants() {
   ...
}
```

`descendants` internally calls `preorder` and ignores `WalEvent::Leave` events.

Reply:
Make sense, thank you